.jenkins/Jenkinsfile

```diff
@@ -250,227 +198,149 @@ def isCanonical() {
-    checkout scm
-    tool type: 'jdk', name: "jdk_${jdk}_latest"
-    def statusCode = 0, attempt = 1
-    retry(2) {
-      if (attempt > 1) { sleep(60 * attempt) }
-      def logfile = "${JOB_NAME}_${BUILD_NUMBER}_${STAGE_NAME}_jdk${jdk}_${arch}_attempt_${attempt}.log"
-      sh "mkdir -p build"
-      def build_dir = sh( returnStdout: true, script:"mktemp -d ${WORKSPACE}/build/build_docker_run.XXXXXXXXXX" ).trim()
-      def build_dir_rel = build_dir - "${WORKSPACE}/"
-      try {
-        if ("jar" != STAGE_NAME) {
-          dir("${build_dir}") {
-            unstash name: "${arch}_${jdk}"
-          }
+def test(command, cell) {
+  def splits = command.splits ? command.splits : 1 
+  def python = cell.python
+  def cython = cell.cython
+  node(getNodeArch(cell.arch)) {
+    withEnv(cell.collect { k, v -> "${k}=${v}" }) {
+      ws("workspace/${JOB_NAME}/${BUILD_NUMBER}/${cell.step}/${cell.arch}/jdk-${cell.jdk}/python-${cell.python}") {
+        cleanWs(disableDeferredWipeout: true)
+        fetchSource(cell.step, cell.arch, cell.jdk)
+        def logfile = "${JOB_NAME}_${BUILD_NUMBER}_${cell.step}_jdk${cell.jdk}_python_${cell.python}_${cell.cython}_${cell.arch}.log"
+
+        def cassandra_dtest_dir = "${WORKSPACE}/build/cassandra-dtest"
+        if (command.containsKey('dtest') && command['dtest']) {
+          checkout changelog: false, poll: false, scm: scmGit(branches: [[name: 'trunk']], extensions: [cloneOption(depth: 1, noTags: true, reference: '', shallow: true), [$class: 'RelativeTargetDirectory', relativeTargetDir: "${WORKSPACE}/build/cassandra-dtest"]], userRemoteConfigs: [[url: 'https://github.com/apache/cassandra-dtest']])
         }
-        echo "Executing ${STAGE_NAME} `build_dir=${build_dir} ${build_script} ${jdk} 2>&1 | tee ${logfile}` on ${NODE_NAME}"
-        statusCode = sh returnStatus:true, script:"build_dir=\"${build_dir}\" ${build_script} ${jdk} 2>&1 | tee ${build_dir}/${logfile}"
 
-        if ("jar" == STAGE_NAME) {
-          dir("${build_dir}") {
-            // only stash the project built files. all dependency libraries are restored from the local maven repo using `ant resolver-dist-lib`
-            stash name: "${arch}_${jdk}"//, includes: "*.jar,classes/**,test/classes/**,tools/**"
+  // sh "mkdir -p build"
+  // def build_dir = sh( returnStdout: true, script:"mktemp -d ${WORKSPACE}/build/build_docker_run.XXXXXXXXXX" ).trim()
+  // def build_dir_rel = build_dir - "${WORKSPACE}/"
+
+  // dir("${build_dir}") {
+    // fetchSource(cell.step, arch, jdk)
+  // }
+        def stageError = true
+        catchError(buildResult: null, message: 'Tests failed', stageResult: 'FAILURE') {
+          sh label: "RUNNING TESTS...", script: "#!/usr/bin/bash \nset -o pipefail;  DEBUG=true python_version='${cell.python}' cython='${cell.cython}' build_dir='${WORKSPACE}/build' cassandra_dtest_dir='${cassandra_dtest_dir}' .build/docker/run-tests.sh ${cell.step} '${cell.split}/${splits}' ${cell.jdk} 2>&1 | tee ${logfile}"
```

Comment:
note, a script failure != test failures.

SUCCESS : scripted exited 0 and all tests passed
UNSTABLE : script exited 0 but there were test failures
FAILURE : script exited 1 (not related to test failures)


It should be the `junit` publisher that changes the stageResult to UNSTABLE for us.